notebooks/interactions_ols.clj

```diff
@@ -0,0 +1,98 @@
+(ns interactions-ols
+  (:require [tablecloth.api :as tc]
+            [tablecloth.pipeline :as tcpipe]
+            [fastmath.stats :as fmstats]
+            [tech.v3.datatype.functional :as fun]
+            [scicloj.metamorph.core :as mm]
+            [scicloj.metamorph.ml :as mm.ml]
+            [scicloj.metamorph.ml.loss :as loss]
+            [scicloj.ml.smile.regression :as regression]
+            [tech.v3.dataset.metamorph :as tmd.mm]
+            [scicloj.kindly.v4.api :as kindly]
+            [scicloj.kindly.v4.kind :as kind]))
+
+^:kindly/hide-code
+(def md
+  (comp kindly/hide-code kind/md))
+
+(md "This examples shows how to do interactions in linear regression with `metamorph.ml`.")
+
+(md "Taking ideas from: [Interaction Effect in Multiple Regression: Essentials](http://www.sthda.com/english/articles/40-regression-analysis/164-interaction-effect-in-multiple-regression-essentials/) by Alboukadel Kassambara")
+
+(md "First we load the data:")
+(def marketing
+  (tc/dataset "https://github.com/scicloj/datarium-CSV/raw/main/data/marketing.csv.gz"
+              {:key-fn keyword}))
+
+(md "## Additive model")
+
+(md "First we build an additive model, which model equation is
+$$sales = b0 + b1 * youtube + b2 * facebook$$")
+
+(def additive-pipeline
+  (mm/pipeline
+   (tmd.mm/set-inference-target :sales)
+   (tcpipe/drop-columns [:newspaper])
+   {:metamorph/id :model}
+   (mm.ml/model {:model-type :smile.regression/ordinary-least-square})))
+
+
+(md "We evaluate it, ")
+(def evaluations
+  (mm.ml/evaluate-pipelines
+   [additive-pipeline]
+   (tc/split->seq marketing :holdout)
+   loss/rmse
+   :loss
+   {:other-metrices [{:name :r2
+                      :metric-fn fmstats/r2-determination}]}))
+
+
+(md "and print the result:")
+(-> evaluations flatten first :fit-ctx :model mm.ml/thaw-model)
+
+(md "We have the following metrics:")
+(md "$RMSE$")
+(-> evaluations flatten first :test-transform :metric)
+
+(md "$R^2$")
+(-> evaluations flatten first :test-transform :other-metrices first :metric)
+
+(md "## Interaction effects")
+(md "Now we add interaction effects to it, resulting in this model equation:
+$$sales = b0 + b1 * youtube + b2 * facebook + b3 * (youtube * facebook)$$")
+(def pipe-interaction
+  (mm/pipeline
+   (tcpipe/drop-columns [:newspaper])
+   (tcpipe/add-column :youtube*facebook (fn [ds] (fun/* (ds :youtube) (ds :facebook))))
+   (tmd.mm/set-inference-target :sales)
+   {:metamorph/id :model}(mm.ml/model {:model-type :smile.regression/ordinary-least-square})))
+
+(md "Again we evaluate the model,")
+(def evaluations
+  (mm.ml/evaluate-pipelines
+   [pipe-interaction]
```

Comment:
You have a single pipe-fn, his means using the pipeline concept has no advantage.
Specially in this case of "regression", we would likely never do AutoML.

Very often in regression, we don't want to find simply teh "Best model", bu we want to learn something about the underlying data / phenomena.

Reply:
But your comment regarding test / train split is correct as well.
The syntax with `ml/evaluate` and pipelines is a bit more compact and 
avoids  some `def`s